packages/jellyfish-transaction-builder/src/txn/txn_builder.ts

```diff
@@ -98,15 +100,21 @@ export abstract class P2WPKHTxnBuilder {
    * @param {OP_DEFI_TX} opDeFiTx to create
    * @param {Script} changeScript to send unspent to after deducting the fees
    * @param {BigNumber} [outValue=0] for the opDeFiTx, usually always be 0.
+   * @param {Array<{ vin: Vin, vout: Vout, prevout: Prevout }>} [customVinVout = []] for custom vin and vout
    */
   async createDeFiTx (
     opDeFiTx: OP_DEFI_TX,
     changeScript: Script,
-    outValue: BigNumber = new BigNumber('0')
+    outValue: BigNumber = new BigNumber('0'),
+    customVinVout: Array<{ vin: Vin, vout: Vout, prevout: Prevout }> = []
```

Comment:
This overloader concept isn't really applicable and it doesn't gel well.

- Given that can inject any `Vin`, the `Vin` can already be collected by `await this.collectPrevouts(minFee)` and included in vins, giving you an invalid Transaction due to duplicated Vins.
- The assumption is with the `Vin` provided that you have the signing rights to it.
- That `{Vin, Vout, Prevout}` comes in a Set, that's not true since only `Vin` is tied to `Prevout`. The amount `Vin` to `Vout` ratio is not exclusive. So an array won't make sense.

Reply:
We also have to ensure collectPrevout doesn't collect the `Vin` you need to sign in a special way.

Reply:
Your setup requires:
1. A custom "Prevouts Collector" implementation since you need control of the `Vins`
2. Injection of `Vout` in a special order.

Reply:
Would it make sense for the upstream implementation to just prepare their Transaction from scratch?

We change `protected async sign` to `public sync sign` to allow that. And anything else that is required we make it public or create an abstraction for it.

Reply:
Hmm, a little lost 😓 correct me if I'm wrong:
- I will implement a "Prevouts Collector" as a method in `P2WPKHTxnBuilder` class
- then combine the outputs of it along with outputs of `await this.collectPrevouts(minFee)` so Vin and Vout can be placed in the correct order that doesn't have duplicates?

Reply:
> prepare their Transaction from scratch

do you mean build the transaction from scratch with vin and vout and then sign it, all in the updatemasternode function in `packages/jellyfish-transaction-builder/src/txn/txn_builder_masternode.ts`?

<img width="483" alt="image" src="https://user-images.githubusercontent.com/35588098/204991989-2ea18efb-568e-44e1-94f7-f37d34be999f.png">

Reply:
Yup, that's right. Since your requirements require vin and vout to be arranged in a certain way. Nobody else will probably need it. It is better to keep the custom logic there instead of bleeding it into an abstraction.

Reply:
Ok, that makes sense. I will do it this way since no other functions need it 👍